portability-spi-cloud/src/main/java/org/dataportabilityproject/spi/cloud/types/PortabilityJob.java

```diff
@@ -35,182 +36,149 @@
+    public Builder setAndValidateJobAuthorization(JobAuthorization jobAuthorization) {
+      switch (jobAuthorization.state()) {
+        case INITIAL:
+        case CREDS_AVAILABLE:
+          // SessionKey required to create a job
+          isSet(jobAuthorization.encryptedSessionKey());
+          isUnset(jobAuthorization.encryptedExportAuthData(),
+              jobAuthorization.encryptedImportAuthData(),
+              jobAuthorization.encryptedPublicKey(),
+              jobAuthorization.encryptedPrivateKey());
+          break;
+        case CREDS_ENCRYPTION_KEY_GENERATED:
+          // Expected associated keys from the assigned worker to be present
+          isSet(jobAuthorization.encryptedSessionKey(),
+              jobAuthorization.encryptedPublicKey(),
+              jobAuthorization.encryptedPrivateKey());
+          isUnset(jobAuthorization.encryptedExportAuthData(),
+              jobAuthorization.encryptedImportAuthData()
+          );
+          break;
+        case CREDS_ENCRYPTED:
+          // Expected all fields set
+          isSet(jobAuthorization.encryptedSessionKey(),
+              jobAuthorization.encryptedPublicKey(),
+              jobAuthorization.encryptedPrivateKey(),
+              jobAuthorization.encryptedExportAuthData(),
+              jobAuthorization.encryptedImportAuthData());
+          break;
+      }
+      return setJobAuthorization(jobAuthorization);
     }
-    this.jobAuthorization = jobAuthorization;
+
+    // For internal use only; clients should use setAndValidateJobAuthorization
+    protected abstract Builder setJobAuthorization(JobAuthorization jobAuthorization);
   }
 
   public Map<String, Object> toMap() {
-    ImmutableMap.Builder<String, Object> builder = ImmutableMap.builder();
-    if (!Strings.isNullOrEmpty(transferDataType)) {
-      builder.put(DATA_TYPE_KEY, transferDataType);
-    }
-    if (!Strings.isNullOrEmpty(exportService)){
-      builder.put(EXPORT_SERVICE_KEY, exportService);
-    }
-    if (!Strings.isNullOrEmpty(importService)){
-      builder.put(IMPORT_SERVICE_KEY, importService);
-    }
-    if (null != jobAuthorization.getEncryptedExportAuthData()) {
-      builder.put(EXPORT_ENCRYPTED_CREDS_KEY, jobAuthorization.getEncryptedExportAuthData());
-    }
-    if (null != jobAuthorization.getEncryptedImportAuthData()) {
-      builder.put(IMPORT_ENCRYPTED_CREDS_KEY, jobAuthorization.getEncryptedImportAuthData());
-    }
-    if (null != jobAuthorization.getEncryptedSessionKey()) {
-      builder.put(ENCRYPTED_SESSION_KEY, jobAuthorization.getEncryptedSessionKey());
+    ImmutableMap.Builder<String, Object> builder = ImmutableMap.<String, Object> builder()
+        .put(DATA_TYPE_KEY, transferDataType())
+        .put(EXPORT_SERVICE_KEY, exportService())
+        .put(IMPORT_SERVICE_KEY, importService())
+        .put(AUTHORIZATION_STATE, jobAuthorization().state().toString())
+        .put(ENCRYPTED_SESSION_KEY, jobAuthorization().encryptedSessionKey());
```

Comment:
Does AutoValue check if these are null/empty too?

Reply:
Good point that I hadn't thought about- AutoValue guarantees the fields are non-null (since no @Nullable annotation) but does not check for empty. But, there should be no harm in persisting an empty value in the map.

Reply:
Ack that makes sense! Thanks!